WRO-735: Add guide document for ui and screenshot tests when contributing (#3043)
enactjs/enact: docs/contributing/ui-testing.md

---

> title: Ui Testing

Reviewer: UI Testing

---

> ## What is UI testing?

Reviewer: UI Testing. Let's be consistent with Unita Testing and Screenshot Testing

> Manual tests are time-consuming and sometimes error occurs. A more efficient approach is to create UI tests so that user tasks are executed in an automated test. With UI test, tests can be executed quickly and reliably in a repeatable way. Simply stated, first priority considering is two things.

Reviewer: Enact supports ui-test-utils package which includes the common WebDriver configurations and some utility modules for executing automated UI tests. With UI test, you can execute user interaction testing quickly and reliably.

* How component handles user actions performed via input devices.

> * Whether elements correctly work as intended.

Reviewer: Not sure

## Prerequisites
Clone sandstone from GitHub, install dependencies and connect the modules using Lerna:

```shell
# clone the repo!
git clone git@github.com:enactjs/sandstone.git
# move in
cd sandstone
# we're using git flow so develop is our working branch
git checkout develop
# install lerna
npm install
```

> ## Creating tests

Reviewer: Creating Tests

Within the UI Library, create an app for testing in `./tests/ui/apps` and create a corresponding test in `./tests/ui/specs`.

	+ src
		+ test
			+ ui
				+ apps
					+testComponent
						testComponent-View.js		<-- create app here
				+ specs
					+ testComponent
						testComponent-specs.js		<-- create spec file here

>  In UI test, create apps to test view and specs. Please refer sample code.

Reviewer: confused to match with the example above

> ---

Reviewer 2: I hope that the link below will also be included in somewhere of this document.
https://github.com/enactjs/ui-test-utils

title: Ui Testing
---
## What is UI testing?
Manual tests are time-consuming and sometimes error occurs. A more efficient approach is to create UI tests so that user tasks are executed in an automated test. With UI test, tests can be executed quickly and reliably in a repeatable way. Simply stated, first priority considering is two things.
* how component handles user actions performed via input devices.
* whether elements correctly work as intended.

## Prerequisites
Clone sandstone from GitHub, install dependencies and connect the modules using Lerna:

```shell
# clone the repo!
git clone git@github.com:enactjs/sandstone.git
# move in
cd sandstone
# we're using git flow so develop is our working branch
git checkout develop
# install lerna
npm install

> # link dependencies
> enact link

Reviewer 2: ```suggestion
```

Author: Suggested Change.

	+ src
		+ test
			+ ui
				+ apps
					+testComponent
						testComponent-View.js		<-- create app here
				+ specs
					+ testComponent
						testComponent-specs.js		<-- create spec file here
 In UI test, create apps to test view and specs. Please refer sample code.
* Button.js

```JS
import Button from '../../../../Button';
import ThemeDecorator from '../../../../ThemeDecorator';

const app = (props) => <div {...props}>
	<div>
		<Button
			id="button"
		>
			Default Button
		</Button>
	</div>
</div>;

export default ThemeDecorator(app);
```

* Button-specs.js

```JS
const Page = require('@enact/ui-test-utils/utils');

describe('Button', function () {

	beforeEach(async function () {
		await Page.open();
	});

	describe('5-way', function () {
		it('should focus disabled button on 5-way right', async function () {
			await Page.spotlightDown();
			await Page.spotlightSelect();
		});
	});
});
```

> The Page component from `@enact/ui-test-utils/utils/` Page contains useful methods for loading tests.

Reviewer 2: ```suggestion
The Page component from `@enact/ui-test-utils/utils/` Page contains useful methods for writing tests.
```

I don't know what is mean of loading.

Author: I removed this part and added link for ui-test-utils.

## Running Tests
For a single-run, execute `npm run test-ui`. 

* Filtering UI by Component:

	`npm run test-ui -- --spec <name-of-component>`

* Re-run tests without building:

	`npm run test-ui -- --skip-build`

* Running with visible browser:

	`npm run test-ui -- --visible`

* Running with filtering by component and without building:

	`npm run test-ui -- --spec <name-of-component> --skip-build`

## Viewing Test Results
By default, Test result display on console whether pass or fail.
If fail occurs, fail log display in console and you can see the screenshot about fail in `./test/ui/errorShots/`.

> Also, you can see the view you made on chrome. This requires that a server be running on port 5000. If you have globally installed the serve command with npm install -g serve you can start the server like this `serve ./test/ui/dist/`.

Reviewer 2: ```suggestion
Also, you can see the view you made on chrome. This requires that a server be running on port 5000. If you have globally installed the serve command with `npm install -g serve` you can start the server like this `serve ./test/ui/dist/`.
```

> To open a specific test app, open the URL path for the test. A specific app display as directory in this page. You can see the app as click the app you wanted.

Reviewer 2: ```suggestion
To open a specific test app, open the URL path for the test. A specific app is displayed as a directory on this page. You can see the app as click the app you wanted.
```

> ## Goal of Ui Testing

Reviewer 2: It would be better if it is at the top like a screenshot test.

---
title: Ui Testing
---
## What is UI testing?
Manual tests are time-consuming and sometimes error occurs. A more efficient approach is to create UI tests so that user tasks are executed in an automated test. With UI test, tests can be executed quickly and reliably in a repeatable way. Simply stated, first priority considering is two things.

> * how component handles user actions performed via input devices.
> * whether elements correctly work as intended.

Reviewer 2: ```suggestion
* How component handles user actions performed via input devices.
* Whether elements correctly work as intended.
```

Author: Suggested Change.

});
```

>  For details on the GitHub methods of UI-test, see [ui-test-utils](https://github.com/enactjs/ui-test-utils).

Reviewer 2: Same with ss test

Reviewer 2: Please recheck it

Author: Sorry, I missed it. Thank you!
